---
title: "16S rRNA gene processing and taxonomic assignment on DNA extracted from samples obtained in Oman in 2019"
subtitle: "Source file: OM19_16S_taxa_assignment_DADA2_Silva138_Bayesian.Rmd"
author: "Daniel Nothaft"
date: "`r format(Sys.Date(), '%d %b %Y')`"
output:
  html_document: 
    df_print: paged # omit to disable paged data table output
    css: stylesheet.css # omit if no need for custom stylesheet
    number_sections: yes # change to no for unnumbered sections
    toc: yes # change to no to disable table of contents
    toc_float: true # change to false to keep toc at the top
    toc_depth: 3 # change to specify which headings to include in toc
    code_folding: show # change to hide to hide code by default
editor_options:
  chunk_output_type: inline
---

# Setup

Data processed using the packages [dada2](https://benjjneb.github.io/dada2/index.html) version `r packageVersion("dada2")`.

Following [this tutorial](https://benjjneb.github.io/dada2/tutorial.html) for 16S gene data processing and [this tutorial](https://benjjneb.github.io/dada2/ITS_workflow.html)

Along with the `dada2` library, we also load the `ShortRead` and the `Biostrings` package, which will help to identify and count the primers present on the raw FASTQ sequence files.  

```{r dada2-Library, warning=FALSE, message=FALSE, tidy=TRUE, results='hold',comment= ""}
library(dada2); # processing sequence data
packageVersion("dada2")
library(tidyverse) # manipulating strings and data frames

# global knitting options for automatic saving of all plots as .png and .pdf. Also sets cache directory.
knitr::opts_chunk$set(
  dev = c("png", "pdf"), fig.keep = "all",
  dev.args = list(pdf = list(encoding = "WinAnsi", useDingbats = FALSE)),
  fig.path = file.path("fig_output/", paste0(gsub("\\.[Rr]md", "/", knitr::current_input()))),
  cache.path = file.path("cache/", paste0(gsub("\\.[Rr]md", "/", knitr::current_input())))
)
```

Load in sequence table
```{r load-seq-tab}
seqtab.nochim <- read_rds("data_output/seqtab.nochim.rds")
```

# Assign taxonomy
It is common at this point, especially in 16S/18S/ITS amplicon sequencing, to assign taxonomy to the sequence variants. The DADA2 package provides a native implementation of the naive Bayesian classifier method for this purpose. The assignTaxonomy function takes as input a set of sequences to be classified and a training set of reference sequences with known taxonomy, and outputs taxonomic assignments with at least minBoot bootstrap confidence.

Here, we use the Silva 138 training set.
```{r assign-taxonomy}
taxa_Silva <- assignTaxonomy(seqtab.nochim, "~/Desktop/Boulder/2018_fall/molecular_methods_ebio/Molecular-Methods/dada2/silva_nr_v138_train_set.fa.gz", multithread=TRUE)
```

Extensions: The dada2 package also implements a method to make species level assignments based on exact matching between ASVs and sequenced reference strains. Recent analysis suggests that exact matching (or 100% identity) is the only appropriate way to assign species to 16S gene fragments.

```{r add-species}
taxa_Silva <- addSpecies(taxa_Silva, "/Users/melo.d/Desktop/Boulder/2018_fall/molecular_methods_ebio/Molecular-Methods/dada2/silva_species_assignment_v138.fa.gz")
```

Letâ€™s inspect the taxonomic assignments:
```{r inspect-taxa-assignments}
taxa.print <- taxa_Silva # Removing sequence rownames for display only
rownames(taxa.print) <- NULL
head(taxa.print)
```

Save taxa table
```{r save-taxtab}
write_rds(taxa_Silva, path = "data_output/taxa.rds", compress = "gz")
```